test/gtest/graphapi_opgraph.cpp

```diff
@@ -0,0 +1,122 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2023 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#include <gtest/gtest.h>
+
+#include <miopen/miopen.h>
+
+#include <miopen/graphapi/opgraph.hpp>
+
+namespace gr = miopen::graphapi;
+
+template <bool IsVirtual>
+gr::Tensor makeDummyTensor(std::string_view name)
+{
+    assert(name.size() <= sizeof(int64_t));
+    int64_t id = 0;
+    std::copy_n(name.begin(), sizeof(id), reinterpret_cast<char*>(&id));
```

Comment:
This may try to read `name` past the end of it.
```suggestion
    std::copy_n(name.begin(), name.size(), reinterpret_cast<char*>(&id));
```

Reply:
> This may try to read `name` past the end of it.

May try to write past the `sizeof(id)`.
It must be something like:
```C++
std::copy_n(name.begin(), std::min(name.size(), sizeof(id)), reinterpret_cast<char*>(&id));
```

Reply:
yes, that's what I ended up doing in the next PR. Will bring the fix back here today.

Reply:
[resolved]

Reply:
@CAHEK7 
> May try to write past the `sizeof(id)`.

No, because of `assert(name.size() <= sizeof(int64_t))`. 

The code that relies of assert actually implies:
- The programmer guarantees that the condition will always be met.
- If condition is not met, the program may work incorrectly

If we afraid that this may happen in release builds (i.e. the programmer can not actually guarantee that the condition will always be met), then assert should be changed to MIOPEN_THROW (but please consider performance penalty).

It must be something like:
> 
> ```c++
> std::copy_n(name.begin(), std::min(name.size(), sizeof(id)), reinterpret_cast<char*>(&id));
> ```

It's sad to say that, but this solution may camouflage the actual errors. This code will work properly only if name fits into id. Otherwise the loss of information will happen and that may lead to incorrect results. Thanks!

Reply:
> @CAHEK7
> 
> > May try to write past the `sizeof(id)`.
> 
> No, because of `assert(name.size() <= sizeof(int64_t))`.

@atamazov 
No. There is no assertion in the release. And that's 100% buffer overrun and assertion is not enough to properly handle this case.

Reply:
@CAHEK7 truncation of the name may spoil uniqueness of the name. When you write `std::min(name.size(), sizeof(id))` you implicitly say that truncation is Ok.

Reply:
yes, we are saying that names greater than 8 characters are not allowed.

Reply:
@amberhassaan perhaps let's just fail the test when `(name.size() > sizeof(int64_t))`?

Reply:
agree. Fix coming in #2855: in #6cc9d8c